src/main/scala/lms/thirdparty/cuda/cuda.scala

```diff
@@ -210,17 +210,112 @@ trait CudaOps extends Base with RangeOps with SizeTOps with StackArrayOps with C
     Wrap[(A,B,C,D,E,F,Dim3,Dim3)=>G](__topFun(f, 6, xn => Unwrap(f(Wrap[A](xn(0)), Wrap[B](xn(1)), Wrap[C](xn(2)), Wrap[D](xn(3)), Wrap[E](xn(4)), Wrap[F](xn(5)))), "__global__"))
 
   // When coding kernel functions, we often need some kernel variables
-  def gridDimX = cmacro[Int]("gridDim.x")
-  def gridDimY = cmacro[Int]("gridDim.y")
-  def blockDimX = cmacro[Int]("blockDim.x")
-  def blockDimY = cmacro[Int]("blockDim.y")
-  def blockDimZ = cmacro[Int]("blockDim.z")
-
-  def blockIdxX = cmacro[Int]("blockIdx.x")
-  def blockIdxY = cmacro[Int]("blockIdx.y")
-  def threadIdxX = cmacro[Int]("threadIdx.x")
-  def threadIdxY = cmacro[Int]("threadIdx.y")
-  def threadIdxZ = cmacro[Int]("threadIdx.z")
+  def gridDimX: Rep[Int] = cmacro[Int]("gridDim.x")
+  def gridDimY: Rep[Int] = cmacro[Int]("gridDim.y")
+  def blockDimX: Rep[Int] = cmacro[Int]("blockDim.x")
+  def blockDimY: Rep[Int] = cmacro[Int]("blockDim.y")
+  def blockDimZ: Rep[Int] = cmacro[Int]("blockDim.z")
+
+  def blockIdxX: Rep[Int] = cmacro[Int]("blockIdx.x")
+  def blockIdxY: Rep[Int] = cmacro[Int]("blockIdx.y")
+  def threadIdxX: Rep[Int] = cmacro[Int]("threadIdx.x")
+  def threadIdxY: Rep[Int] = cmacro[Int]("threadIdx.y")
+  def threadIdxZ: Rep[Int] = cmacro[Int]("threadIdx.z")
+
+
+  // Here we will implement some cuda kernel functions using the `cudaGlobalFun`
+
+  def cudaFill[T:Manifest](implicit __pos: SourceContext) = cudaGlobalFun {
+    (data: Rep[Array[T]], value: Rep[T], size: Rep[Int]) =>
+      val stride = gridDimX * blockDimX
+      val tid = threadIdxX + blockIdxX * blockDimX
+      for (i <- tid.until(size, stride): Rep[Range]) {
+        data(i) = value
+      }
+    }
+
+  // cudaCap: cap the absolute value of `data`
+  def cudaCap[T:Numeric:Manifest](implicit __pos: SourceContext) = cudaGlobalFun {
+    (data: Rep[Array[T]], bound: Rep[T], size: Rep[Int]) =>
+      val stride = gridDimX * blockDimX
+      val tid = threadIdxX + blockIdxX * blockDimX
+      for (i <- tid.until(size, stride): Rep[Range]) {
+        __ifThenElse(data(i) > bound, {data(i) = bound}, {})
```

Comment:
use normal `if` syntax here?

Reply:
I tried. It works on ml00 machine but in GitHub Actions, it errors because I have to use `@virtualize`. (something like: the macro is defined and used at the same level)